Implement request filter pipeline

daltonmatos opened the pull request:
Hello @orlandohohmeier,

Here is a implementation candidate for a "filters pipeline" feature. This will enrich plugin even more, as long as the core Marathon UI adds some extension points where it calls the filter pipeline.

Let me know what do you think about this implementation, so I can adjust the code to be able to eventually merge.

Things I didn't like the way it's done, but don't know how to do it differently.

* The constants that define the name of the filters: Don't know if there'a a way to continue to use `PipelineStore.PRE_AJAX_REQUEST` without the need of the constructor.
* The `clearPipeline()` method is only used in the tests. I could not clean the pipeline before each tests without this method.
* Ideally the object that holds all pipelines (here `FILTERS_PIPELINE`) should be visible only to the `PipelineStore` module, and not be an attribute of the `PipelineStore` class, also don't know how to achieve this.


Thanks a lot!

daltonmatos commented:
Hello @orlandohohmeier,

I will adjust some things that I learned how to do (like change `module.exports` to `export const`) and will update the commits. Then you can take a look and start the review.

The overall idea is still the same, however.

Thanks,

daltonmatos commented:
Hello @orlandohohmeier,

the implementation is a little better now and is ready for you to start reviewing.

Thanks,

daltonmatos commented:
Hello @orlandohohmeier,

made some changes and the implementation is getting better. If you could take a look.

About the new instance of the Pipeline passed to every loaded plugin: If you think this is a problem, we can extract this instance to a variable and pass the same Pipeline object to every plugin.

Also, if you think this is better we can extract all hook names to a separated file, for example: `plugin/sdk/stores/hooks.js`.

Let me know what you think.

Thanks,

orlandohohmeier commented:
Hi @daltonmatos, 

Please excuse the late response. I've been on the road for the last two weeks and wasn't able to review your code.

I like the idea of a filter/pipeline mechanism to enrich the plugin system and I'm happy to help move this one forward. We might want to think about ways to ensure that the input and output types are consistent as I've seen such mechanism fail many times due to the lack of type safety and the related set of errors.

Making `PipelineStore` instantiable was a great idea and is the way to go to circumvent the need for a `clearPipeline` method. Using the plugin interface to share a new instance of the pipeline store is fine— think of it as an IOC container.

I'd refactor the pipeline store as follows to address the open issues you've initial mentioned and ensure data integrity. 

```
export default class PipelineStore {

  constructor() {
    /**
     * Private Context
     *
     * @typedef {Object} PipelineStore~Context
     * @property {PipelineStore} instance - Current pipeline store instance
     * @property {Map} pipelines
     */
    const context = {
      instance: this,
      pipelines: new Map()
    };

    this.applyPipeline = this.applyPipeline.bind(context);
    this.registerOperator = this.registerOperator.bind(context);
    this.unregisterOperator = this.unregisterOperator.bind(context);

  }

  /**
   * Apply pipeline
   *
   * @this PipelineStore~Context
   * @param {Symbol|String} id - Pipeline identifier
   * @param {*} input
   * @return {*} output
   */
  applyPipeline(id, input) {
    if (!this.pipelines.has(id)) {
      return input;
    }

    return this.pipelines.get(id)
      .reduce((accumulator, operator) => {
        try {
          return operator(accumulator);
        } catch (error) {
          // We might want to log a proper error to help users debug  
          return accumulator;
        }
      }, input);

  };

  /**
   * @this PipelineStore~Context
   * @param {Symbol|String} id
   * @param {function} operator
   */
  registerOperator(id, operator) {
    if (!this.pipelines.has(id)) {
      this.pipelines.set(id, []);
    }

    const pipeline = this.pipelines.get(id);

    if (!pipeline.includes(operator)) {
      pipeline.push(operator);
    }
  };

  /**
   * @this PipelineStore~Context
   * @param {Symbol|String} id
   * @param {function} operator
   */
  unregisterOperator(id, operator) {
    if (!this.pipelines.has(id)) {
      return;
    }

    const pipeline = this.pipelines.get(id);

    if (pipeline.includes(operator)) {
      pipeline.splice(pipeline.indexOf(operator), 1);
    }
  }
};
```

Creating a binding a private context helps us to ensure data integrity. Please note that it's rather hard to properly extend this class due to the usage of the private context which in this case shouldn't be a problem.

daltonmatos commented on src/js/plugin/sdk/stores/PipelineStore.js:
```diff
@@ -0,0 +1,44 @@
+
+export const FILTERS_PIPELINE = {};
```
Sure! The idea is to let only the `PipelineStore` modify the internal state of the pipelines. Currently exporting this variable makes it possible to other components of the core UI do modify it (that's what I'm currently doing in the test cases) but this is not true for plugins, this is because plugins cannot import the module directly, all that plugins see are the objects exported to its `pluginWindow.marathonPluginInterface`, so no plugin could ever reach the original exported `FILTERS_PIPELINE`, right? Does this make sense?

daltonmatos commented on src/js/plugin/sdk/stores/PipelineStore.js:
```diff
@@ -0,0 +1,44 @@
+export const FILTERS_PIPELINE = {};
+const PRE_AJAX_REQUEST = "PRE_AJAX_REQUEST";
+
+export const Hooks = {
```
Sure! Will do it.

daltonmatos commented:
Hello @orlandohohmeier, 

Thank you very much for your time to review this PR.

> Making PipelineStore instantiable was a great idea and is the way to go to circumvent the 
> need for a clearPipeline method. Using the plugin interface to share a new instance of the pipeline store is fine— think of it as an IOC container.

Indeed this is great, but if we have a new `pipelines` (`new Map()`) object for each new `PipelineStore` the following scenario won't work:

Plugin A registers a `PRE_AJAX_REQUEST` operator;
Plugin B also registers a `PRE_AJAX_REQUEST` operator;

When `ajaxWrapper` calls `pipeline.applyPipeline(PRE_AJAX_REQUEST, options)` the two registered operators (one for each plugin) won't run, because `Plugin A`, `Plugin B` and `ajaxWrapper` have a different instance of `PipelineStore`. Even if we take care of passing the same instance to all plugins, that is, `PluginLoader` instantiates `PipelineStore` only once, `ajaxWrappeer` would still have a different instance, because whoever imports `PipelineStore` must create an instance before using it.

That's why I choosed to have `FILTERS_PIPELINE` as a global object inside the module, so every instance of `PipelineStore` would use it and the tests could prepare it before each testcase.

I don't know how is the best way to solve this. What do you think would be the best way to achieve this behavior: No plugin can mess with the internal pipelines state, but all plugins share the same pipeline internal state.

I thought that having the `FILTERS_PIPELINE` importable only by the core UI code would be acceptable, since we control all of the core UI code and we could agree not to modify the internal state directly and always use `PipelineStore` for this. The only exception would be the unit tests code.

Let me know what do you think.

Thanks a lot.

orlandohohmeier commented:
Hi @daltonmatos,

you're right, with the current proposal we would run into issues as the plugins and marathon UI wouldn't share the same instance of the PipelineStore — I forgot that we don't use the plugin interface to share instance internally as well. 
I'd  rename the current `PipelineStore` to `PipelineModel` and create a PipelineStore which acts as a singleton and ensures that all parties share the same instance of the `PipelineModel` to circumvent the issues. This would allow you to properly test the `PipelineModel` without opening its internals to the world. What do you think?

daltonmatos commented:
Could work but I have some doubts:

Are you suggesting the `PipelineModel` to be an internal **attribute** of `PipelineStore`?

Because if yes, even if they share the same instance of `PipelineModel` (which is the correct behavior), being an internal attribute (`this.pipeline = ...` in the `getInstance()` singleton method, for example) would allow any plugin do modify the internal state of the pipeline.

When I created the `FILTERS_PIPELINE` as a global variable **in the module scope** I was trying to hide this variable from all plugins, since this is not even exported to them. And they can't do anything like:

```
const {
  PipelineStore
} = window.marathonPluginInterface;

PipelineStore.pipeline = {};
```

becasue there is no such attribute inside `PipelineStore` instance, since it uses the module scoped `FILTERS_PIPELINE` object.

If your are not suggesting  the `PipelineModel` to be an internal attribute of `PipelineStore`, I think we return to the same point, that will be one object published to all plugins that uses a module scoped object that stores the internal state of the pipeline, we just have different names: `PipelineStore` using `FILTERS_PIPELINE` or `PipelineStore` using `PipelineModel`.

About the tests on this approach, the test code should still be able to import the instance of `PipelineModel` (thus it would be exportable to the core UI code) because it needs to do two things (when testing `PipelineStore` behavior):

* I have a test that instantiate two `PipelineStore`, adds a filter to each one and then checks that both filters are in the internal pipeline state;
* The tests need to clear the pipeline state so that tests do not interfere one another.

Do you think is it too bad to let an object importable to the core UI code?

Thanks a lot!!

daltonmatos commented:
Hello @orlandohohmeier, any thoughts?

Thanks,

orlandohohmeier commented:
Hi @daltonmatos,

Imaging the pipeline model throws an error – where would you look for bugs? In case your local state is importable to the core UI code you would need to check not only the pipeline model but also the entire core UI to ensure that none of the components mutates that the state by accident. 
Making sure that the state is guarded drastically decreases the scope of errors and makes it easier to reason about the respective implementation. Let's try to work-around the need for an unguarded state even though that many of the old stores leak state. 

I was thinking about the `PipelineStore` being nothing more than a proxy sharing the `PipelineModel` instance as illustrated in the following example:

```js
// PipelineStore.js

export default  new PipelineModel();
```

You could also use `Object.defineProperty` to define a read-only property.

```js
// PipelineStore.js

export default new class {
  constructor() {
    Object.defineProperty(this, "pipelines", {value: new PiplelineModel()});
  }
}
```

Both of these approaches make it _impossible_ for any plugin to tamper with the `PipelineStore` or `PipelineModel`. 

Your test would then test that the `PipelineStore` always provide the same instance of `PipelineModel` and that the `PipelineModel` works properly.

daltonmatos commented:
Hello @orlandohohmeier,

I implemented the a idea of `PipelineStore` being a wrapper for `PipelineModel`. Let me know if this was what you were thinking of.

Also extracted the constants to a separated file. Don't know if the names are ok, but let's focus on the idea and then we can change the namings, if needed.

Let me know what do you think.

Thanks.

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,48 @@
+
+};
+
+export default new PipelineModel();
```
Exporting an instance makes it unnecessary hard to test the `PipelineModel` and renders the store useless. Let's export the class here and use the store to create and provide the `PipelineModel` instance.

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,48 @@
+    this.pipelines.clear();
+  };
+
+  applyFilter(operationName, filterParameters) {
```
I'd rename this method to `applyPipeline` as a pipeline could be used for many things other than filtering a result set otherwise I'd argue that we should restructure the model to prove a predicate that one can apply instead of applying the filter.

daltonmatos replied:
:+1:

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,48 @@
+    this.pipelines.clear();
+  };
+
+  applyFilter(operationName, filterParameters) {
```
We should also document that a pipeline can't handle more than one input parameter (could be an object or list though) as the output of one operation will be the input of the next operation.

daltonmatos replied:
Where is the best place to document this? As a comment in this method?

orlandohohmeier replied:
I'd add a JSDoc comment for now.

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,48 @@
+  applyFilter(operationName, filterParameters) {
+    const pipeline = this.getPipeline(operationName);
+
+    pipeline.forEach((filter) => {
```
Let's rewrite this method to properly pass the input and output parameter so that `applyPipeline(FOO, "bar") === c(b(a("bar"))))` given we've registered the functions `a`, `b`, `c` as operators for "FOO".

```js
function a(input){
 return input.toUpperCase();
}

function b(input){
 return input.split("").reverse().join("");
}

function c(input){
 return input.replace(/[AEUIO]/g,"");
}

pipelines =  new PipelineModel();
pipelines.registerOperator("FOO", a); 
pipelines.registerOperator("FOO", b); 
pipelines.registerOperator("FOO", c);

pipelines.applyPipeline("FOO", "bar") === c(b(a("bar")))
```

https://github.com/mesosphere/marathon-ui/pull/819#issuecomment-352718900 illustrates how to implement such a method.

daltonmatos replied:
:+1:

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,48 @@
+    }
+  };
+
+  getPipeline(pipelineName) {
```
I wouldn't expose this method as theoretically allows unwanted mutations. I'd also argue that as not necessary to extract the respective functionality as we only need to get the pipelines in a very few places and only in one place we want to actually create a new pipeline if there is none.

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineStore.js:
```diff
@@ -0,0 +1,21 @@
+import PipelineModel from "./PipelineModel";
+
+export default class PipelineStore {
```
We should export and instance of `PipelineStore` as the pipeline store will create and manage the `PipelineModel` should be a _singleton_ similar to all of our stores.

orlandohohmeier commented:
Hi @daltonmatos,

I very much like where this is going! The current instance handling seems a little of though and doesn't solve the issues we've discussed. I've started to also comment on other parts of the implementations – feel free to ignore those comments for now. I was also wondering wether you want to implement a general purpose pipeline mechanism which one can use to apply a set of operators or provide a mechanism to filter data using the registered predicates. If you think we only need the latter I'd propose a slightly different implementation and would adjust the naming.

Once again, thanks for pushing the plugin interface.

daltonmatos commented:
Hello @orlandohohmeier,

The idea, for now, is to just have the ability to filter some values in a very simple form. Only one parameter passed to the filters, no priority when running the registered callbacks.

The generic pipeline implementation could be postponed until really needed, which is not the case for now.

Let's do the minimal filtering implementation for now and think about the pipeline (if ever needed) in the future, what do you think?

Let me know what are you suggestions about the naming and the different implementation you mentioned above so we can move on.

Thanks a lot for your time reviewing my submissions. :tada:

daltonmatos commented:
Hello @orlandohohmeier,

What do you think about making this implementation as simple as possible so we can merge it?

Let me know what are the suggestions you mentioned earlier so I can implement them on this PR.

Thanks,

orlandohohmeier commented:
@daltonmatos please excuse that I didn't come back to you earlier. I've tried a few things but considering that you actually want to mutate an input object (e.g. `options`) I'd implement the pipeline as outlined in https://github.com/mesosphere/marathon-ui/pull/819#discussion_r158945625 what do you think?

daltonmatos commented:
Sure @orlandohohmeier, no problem. I will write this implementation.

I will also use the names you suggested: `applyPipeline()`, `registerOpetator()` and `unregisterOperator()`.

Thanks,

daltonmatos commented:
Hello @orlandohohmeier, 

please take a look at the implementation. Most of it was already done, just didn't push yet.

Thanks,

orlandohohmeier commented:
@daltonmatos the last days were super busy week and I haven't had the time to review your code. Will look into it next week. Hope you don't mind. 🙂

daltonmatos commented:
No problem @orlandohohmeier. Let me know when you have some time to review this PR.

daltonmatos commented:
Hello @orlandohohmeier,

have you had any time to review this PR?

Thanks,

orlandohohmeier commented:
Hi @daltonmatos, unfortunately I haven't had the time to look into this. I'm super sorry. Will look into it first thing tomorrow morning.

daltonmatos commented:
No problem, thanks!

orlandohohmeier commented:
@daltonmatos Thanks a lot for your effort to ship this new capability. I've added a few comments here and there including stylistic ones that I flagged with the 🎨  emoji.

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,54 @@
+    this.pipelines = new Map();
+  };
+
+  clearPipeline() {
```
Done.

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,54 @@
+    this.pipelines.clear();
+  };
+
+  applyPipeline(operationName, filterParameters) {
```
Done. Tried to adjust. :smile:

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,54 @@
+         }, filterParameters);
+  };
+
+  registerOperator(operationName, callback) {
```
Done.

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,54 @@
+  deregisterOperator(operationName, callback) {
+    const pipeline = this.getPipeline(operationName);
+
+    var callbckPosition = pipeline.indexOf(callback);
```
Done.

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineOperations.js:
```diff
@@ -0,0 +1,8 @@
+  PRE_AJAX_REQUEST
+};
+
+export default PipelineOperations;
```
Done.

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineStore.js:
```diff
@@ -0,0 +1,23 @@
+  };
+
+  deregisterOperator(operationName, callback) {
+    return this.deregisterOperator(operationName, callback);
```
Done. Thanks. Didn't spot this. Should be `return this.pipeline.deregisterOperator(operationName, operator)`

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+
+const pipelineModel = new PipelineModel();
+
+describe("PipelineModel tests", function () {
```
Done.

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+describe("PipelineModel tests", function () {
+
+    beforeEach(function (){
+      pipelineModel.clearPipeline();
```
Done.

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+      pipelineModel.clearPipeline();
+    });
+
+    it("should return an empty list for a non-existant pipeline", function () {
```
Done.

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+      expect(pipelineModel.getPipeline(PipelineOperations.PRE_AJAX_REQUEST)).to.have.lengthOf(0);
+    });
+
+    it("should clear the pipeline when requested to", function () {
```
Done.

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+      expect(pipelineModel.getPipeline(PipelineOperations.PRE_AJAX_REQUEST)).to.have.lengthOf(0);
+    });
+
+    it("should add a new function to the pipeline", function () {
```
Done.

daltonmatos commented on src/test/units/PipelineModel.test.js:
```diff
@@ -0,0 +1,100 @@
+
+    it("should call all functions of the chosen pipeline", function () {
+      var filter_one = function(data) {
+        data.value = data.value + "_filter_one";
```
Done.

daltonmatos commented on src/test/units/PipelineStore.test.js:
```diff
@@ -0,0 +1,25 @@
+import {expect} from "chai";
+import PipelineStore from "../../js/plugin/sdk/pipeline/PipelineStore";
+import pipelineModel from "../../js/plugin/sdk/pipeline/PipelineModel";
```
Done.

daltonmatos commented on src/test/units/PipelineStore.test.js:
```diff
@@ -0,0 +1,25 @@
+      const filter = function (data) {return data;};
+      const filter2 = function (data) {return data;};
+
+      const pipeline_one = PipelineStore;
```
Done. Don't know if this suite makes sense at all, I can't import `PipelineStore` twice so I can't even simulate two instance of `Pipelinestore`. What do you think about removing this suite completely?

About the forwarding test, I would just register two operators and check that they ran, this would confirm that `PipelineStore` works as expected. What do you think?

orlandohohmeier replied:
You should be able to import the store as follows:

```js
import {
  default as TestPipelineStore
} from "../../js/plugin/sdk/pipeline/PipelineStore";

// ...

PipelineStore.pipelines === TestPipelineStore.pipelines;
```

I'd use `sinon` to check wether the respective functions were called as illustrated in the following:

```js
sinon.spy(PipelineStore.pipeline, "applyPipeline");

expect(PipelineStore.pipeline.applyPipeline).to.be.true;
```

daltonmatos commented:
All done @orlandohohmeier. Please Take a look.

The suite for the `PipelineStore` is still broken, as I would like to know what should we do with it.

Also let me know if the publushing of `PRE_AJAX_REQUEST` to the plugins (inside `PluginLoader.js`) is correct.

awaiting for your response/review.

Thanks a lot!

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineStore.js:
```diff
@@ -0,0 +1,23 @@
+class PipelineStore {
+
+  constructor() {
+    Object.defineProperty(this, "pipeline", {value: new PipelineModel()});
```
🎨  I'd rename the value to "pipelines" as we're allowing users to register operators for multiple different pipelines.

orlandohohmeier commented:
@daltonmatos  thanks for addressing the comments. 💜  Found a few more minor things but those aside the code looks good.

daltonmatos commented:
All Done @orlandohohmeier,

let me know what you think. I also implemented a ran locally an UI plugin that uses this new interface. Works as expected.

Thanks,

daltonmatos commented on src/js/plugin/sdk/pipeline/PipelineStore.js:
```diff
@@ -0,0 +1,23 @@
+import PipelineModel from "./PipelineModel";
+
+class PipelineStore {
```
Makes sense, but I would leave it inside the `plugin/sdk/stores` folder and not in `js/stores` folder as I think the pipeline is feature provided by the plugin SDK (yet to be extracted from the core source-code). The UI core can use it, of couse, but importing from `plugin/sdk/store/PipelineStore`.

daltonmatos commented:
Hello @orlandohohmeier,

any thoughts about this? I implemented all your suggested reviews.

Thanks,

orlandohohmeier commented on src/js/plugin/sdk/pipeline/PipelineModel.js:
```diff
@@ -0,0 +1,46 @@
+      return filterParameters;
+    }
+    return this.pipelines.get(pipelineName)
+      .reduce((accumulator, pipelineCallback) => {
```
🎨  I'd rename `pipelineCallback` to `operator` to stick with the general naming scheme.

daltonmatos replied:
Done.

orlandohohmeier commented:
@daltonmatos thanks a lot for your relentless work on this one. This is a great addition which will enable new plugins. Well done!  💯  

Please squash your commits before we merge this one, as not all of them work independently of any later commits, and pass the linter as well as the test suite. Thanks!

daltonmatos commented:
Great @orlandohohmeier,

Do you mind if I squash all commit into one single commit? This implementation is overall small and simple, so I think i will still be easy to review/understand even if contained in a single commit.

Let me know and I will update this PR and it will be ready do be merged.

Thanks,

daltonmatos commented:
Hello @orlandohohmeier,

I squashed all commits into one single commit. Thought this was easier than trying to pick some commits to squash. If you prefer, I still have all original commits, if you think it's necessary I can push them back here in this PR and we can choose which commits to squash.

Thanks **a lot** for your reviews and for accepting this code change.

orlandohohmeier commented:
@daltonmatos thanks for squashing your commits 🚀
